### PR TITLE
feat(containers): Add SAS and SAS Studio IDE

### LINKED ISSUE #330: Add SAS and SAS Studio IDE
# Description

Addition of SAS for CLI, SASPY for Notebooks, and SAS Studio (Web IDE).

> **Note**: I use another base container to grab the SAS F.S. where as the base container grabbed the SAS F.S. from an azure share but in order to not leak credentials this was the best method and also faster in CI due to docker compression.

> **Note**: SAS alone is about 6GB so the full container image is 12.0GB at the moment.

# Remaining Tasks

- [x] Using custom extension of jupyter-server-proxy add a button which will allow for a user to launch SAS Studio IDE
- [x] Resolve authorization issue by correctly setting cookie for SAS Studio IDE using the custom extension
- [ ] Add CI step
- [x] Add GateKeeper Policy to restrict to StatCan employees

# Custom Extension

All the custom extension needs to do is call the following script when the button is clicked:

```sh
/usr/local/SASHome/studioconfig/sasstudio.sh start
```

Along with forwarding to the correct address:

```
http://127.0.0.1:8888/notebook/username/notebookname/proxy/38080/SASStudio/
```

## Quick Documentation

> **Note**: Reminder that important files can be saved in your Azure Files remote mount.

### Environment Launch Screen

Eventually the custom image will appear in the drop down on the preceding screen but for now this is the post selection screen showing that we selected the SAS container and gave it 1 CPU and 4 GB memory.

01-select-environment

02-launch-screen

### SASPY with Notebooks

SAS Kernel through SASpy allows a python process to connect to SAS 9.4 and run SAS code, generated by the supplied object and methods or explicitly user written, and returns results as text, HTML5 documents (via SAS ODS), or as Pandas Data Frames. It supports running analytics and returning the resulting graphics and result data. It can convert between SAS Data Sets and Pandas Data Frames. It has multiple access methods which allow it to connect to local or remote Linux SAS, IOM SAS on Windows or Linux (Including Grid Manager), and local PC SAS. It can run w/in Jupyter Notebooks, in line mode python or in python batch scripts.

```sh
title "Stock High, Low, and Close";

proc sgplot data=sashelp.stocks;
  where Date >= '01JAN2005'd and stock='IBM';
  highlow x=date high=high low=low / close=close;
run;

title;
```

03-sas-notebook

### SAS Studio

We also added the option for launching SAS Studio all accomplished by just clicking the button in the launch screen.

Please note it is a little slow to start up but this is normal though for performance might need to give more then 1 CPU.

04-sas-studio

## Links

- [SASPY](https://github.com/sassoftware/saspy)
- [SAS for Containers](https://www.sas.com/en_us/solutions/cloud/for-containers.html)
- [Whitepaper](https://www.sas.com/content/dam/SAS/documents/briefs/solution-brief/en/sas-for-containers-109970.pdf)
- [Technical Paper](https://support.sas.com/content/dam/SAS/support/en/technical-papers/understanding-containers-sas-9-4-container-deployment.pdf)

### PR DESCRIPTION
**Pull request by sylus**
Signed-off-by: sylus <sylus1984@gmail.com>

This resolves #330

# Description

This P.R. will add SAS, SASPY for Notebooks, and SAS Studio (Web IDE).

# Tests / Quality Checks

## Automated Testing/build and deployment
- [ ] Does the image pass CI successfully (build, pass vulnerability scan, and pass automated test suite)?
- [ ] If new features are added (new image, new binary, etc), have new automated tests been added to cover these?
- [ ] If new features are added that require in-cluster testing (e.g. a new feature that needs to interact with kubernetes), have you added the `auto-deploy` tag to the PR before pushing in order to build and push the image to ACR so you can test it in cluster as a custom image?

## JupyterLab extensions

- [ ] Are all extensions "enabled" (`jupyter labextension list` from inside the notebook)?

## VS Code tests

- [ ] Does VS Code open?
- [ ] Can you install extensions?

## Code review

- [ ] Have you added the `auto-deploy` tag to your PR before your most recent push to this repo?  This causes CI to build the image and push to our ACR, letting reviewers access the built image without having to create it themselves
- [ ] Have you chosen a reviewer, attached them as a reviewer to this PR, and messaged them with the SHA-pinned image name for the final image to test (e.g. `k8scc01covidacr.azurecr.io/machine-learning-notebook-cpu:746d058e2f37e004da5ca483d121bfb9e0545f2b`)?

### CODE REVIEW

**Comment by blairdrummond** [comment]
@sylus is there a way for us to control whether or not the notebook is able to start SAS? I think we want to prevent external users from using SAS (correct me if I'm mistaken @brendangadd), so I think we may want to inject the token (or whatever SAS uses) at runtime, even if it's static

**Comment by zachomedia** [comment]
From my understanding, I don't think so. The license stuff is all done at installation time, so you'll probably just have to use a Gatekeeper policy for that.

File: output/sas/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+
+COPY sascfg.py /opt/conda/lib/python3.8/site-packages/saspy/sascfg.py
+
+RUN jupyter nbextension install --py sas_kernel.showSASLog && \
+    jupyter nbextension enable sas_kernel.showSASLog --py && \
+    jupyter nbextension install --py sas_kernel.theme && \
+    jupyter nbextension enable sas_kernel.theme --py && \
+    jupyter nbextension list
+
+RUN pip install git+https://git.zacharyseguin.ca/temp/jupyter-sasstudio-proxy.git
```
Comment (zachomedia) [inline output/sas/Dockerfile:318]: Just putting a @TODO: This needs to be moved the StatCan GitHub Org.

**Comment by blairdrummond** [comment]
@zachomedia can we use Manifest annotations for this?

https://github.com/opencontainers/image-spec/blob/main/annotations.md#pre-defined-annotation-keys

I would be happier if our gatekeeper policies where selecting/rejecting based on proper metadata, rather than just using image names

File: output/sas/Dockerfile (hunk 1 above)
Comment (sylus) [reply output/sas/Dockerfile:318]: Does this need a new repo @zachomedia or can it be inlined in the resources/sas folder?

Comment (sylus) [reply output/sas/Dockerfile]: I ended up inlining this and all works :)

File: output/sas/jupyter-sasstudio-proxy/jupyter_sasstudio_proxy/__init__.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+logger = logging.getLogger(__name__)
+logger.setLevel("INFO")
+
+def setup_sasstudio():
+  def _get_cmd():
+    return [
+      "bash",
+      "-c",
+      "/usr/local/SASHome/studioconfig/sasstudio.sh start && cat"
```
Comment (blairdrummond) [inline output/sas/jupyter-sasstudio-proxy/jupyter_sasstudio_proxy/__init__.py:12]: @sylus Do you know if it's possible to run the server at startup automatically? (I.e. have it pre-warmed for once users press this button). The launch time is quite long for SAS Studio, so it'd make the feel more snappy

**Review by blairdrummond — DISMISSED** [review]
LGTM!

**Pull request merged**